presto-cli/src/main/java/com/facebook/presto/cli/StatusPrinter.java

```diff
@@ -443,4 +453,103 @@ private static double percentage(double count, double total)
+        private final long processedRows;
+
+        public ProgressStats(long timestampNanos, long processedBytes, long processedRows)
+        {
+            this.timestampNanos = timestampNanos;
+            this.processedBytes = processedBytes;
+            this.processedRows = processedRows;
+        }
+
+        public long getTimestampNanos()
+        {
+            return timestampNanos;
+        }
+
+        public long getProcessedBytes()
+        {
+            return processedBytes;
+        }
+
+        public long getProcessedRows()
+        {
+            return processedRows;
+        }
+    }
+
+    private class ProgressStatsHistory
+    {
+        private final Map<String, LinkedList<ProgressStats>> stageProgressStatsHistory = new HashMap<>();
+        private final LinkedList<ProgressStats> statementProgressStatsHistory = new LinkedList<>();
+        private final Duration statsHistoryInterval = new Duration(60, SECONDS);
+        private final long start;
+
+        public ProgressStatsHistory(long timestampNanos)
+        {
+            this.start = timestampNanos;
+            this.statementProgressStatsHistory.add(new ProgressStats(start, 0, 0));
+        }
+
+        public void addStatementProgressStats(ProgressStats progressStats)
+        {
+            statementProgressStatsHistory.add(progressStats);
+            popOldStats(statementProgressStatsHistory);
+        }
+
+        public String getStatementInstantDataRate()
+        {
+            return getDataRate(statementProgressStatsHistory);
+        }
+
+        public String getStatementInstantRowCountRate()
+        {
+            return getRowCountRate(statementProgressStatsHistory);
+        }
+
+        public void addStageProgressStats(String id, ProgressStats progressStats)
+        {
+            if (!stageProgressStatsHistory.containsKey(id)) {
+                stageProgressStatsHistory.put(id, new LinkedList<>());
+                stageProgressStatsHistory.get(id).add(new ProgressStats(start, 0, 0));
+            }
+            LinkedList<ProgressStats> stageHistory = stageProgressStatsHistory.get(id);
+            stageHistory.add(progressStats);
+            popOldStats(stageHistory);
+        }
+
+        public String getStageInstantDataRate(String id)
+        {
+            return getDataRate(stageProgressStatsHistory.get(id));
+        }
+
+        public String getStageInstantRowCountRate(String id)
+        {
+            return getRowCountRate(stageProgressStatsHistory.get(id));
+        }
+
+        private void popOldStats(LinkedList<ProgressStats> progressStatsHistory)
+        {
+            while (succinctNanos(progressStatsHistory.getLast().getTimestampNanos() - progressStatsHistory.getFirst().getTimestampNanos()).compareTo(statsHistoryInterval) > 0) {
+                progressStatsHistory.removeFirst();
+            }
+        }
+
+        private String getDataRate(LinkedList<ProgressStats> progressStatsHistory)
```

Comment:
- Can we make this method static? Same for `getRowCountRate`
- Since the formatted data rate (string) is returned. Make the method name `getFormattedDataRate`. `getDataRate` is expected to return a double, or a (DataSize, Duration) pair :).